tidy3d/components/device/source.py

```diff
@@ -0,0 +1,94 @@
+"""Defines heat material specifications"""
+from __future__ import annotations
+
+from abc import ABC
+from typing import Union, Tuple
+
+import pydantic.v1 as pd
+
+from .viz import plot_params_heat_source
+
+from ..base import cached_property
+from ..base_sim.source import AbstractSource
+from ..data.data_array import TimeDataArray
+from ..viz import PlotParams
+
+from ...constants import VOLUMETRIC_HEAT_RATE
+from ...exceptions import SetupError
+from ...log import log
+
+
+class DeviceSource(AbstractSource, ABC):
+    """Abstract device source."""
+
+    structures: Tuple[str, ...] = pd.Field(
+        title="Target Structures",
+        description="Names of structures where to apply heat source.",
+    )
+
+    @cached_property
+    def plot_params(self) -> PlotParams:
+        """Default parameters for plotting a Source object."""
+        return plot_params_heat_source
+
+    @pd.validator("structures", always=True)
+    def check_non_empty_structures(cls, val):
+        """Error if source doesn't point at any structures."""
+        if len(val) == 0:
+            raise SetupError("List of structures for heat source is empty.")
+
+        return val
+
+
+class HeatSource(DeviceSource):
+    """Volumetric heat source.
+
+    Example
+    -------
+    >>> heat_source = HeatSource(rate=1, structures=["box"])
+    >>> # analytical expression as heat source
+    >>> expression = f"{r}*(sqrt(x) + sin(y)*pow(z,3))"
+    >>> heat_source_expr = HeatSource(rate=expression, structures=["box"])
+    """
+
+    rate: Union[str, float, TimeDataArray] = pd.Field(
+        title="Volumetric Heat Rate",
+        description="Volumetric rate of heating or cooling (if negative) in units of "
+        f"{VOLUMETRIC_HEAT_RATE}.",
+        units=VOLUMETRIC_HEAT_RATE,
+    )
+
+
+class HeatFromElectricSource(DeviceSource):
+    """Volumetric heat source generated from an electric simulation.
+    If a `HeatFromElectricSource` is specified as a source, appropriate boundary
+    conditions for an electric simulation must be provided, since such a simulation
+    will be executed before the heat simulation can run.
+
+    Example
+    -------
+    >>> heat_source = HeatFromElectricSource(structures=["box"])
+    """
+
+
+class UniformHeatSource(HeatSource):
+    """Volumetric heat source.
+
+    Example
+    -------
+    >>> heat_source = UniformHeatSource(rate=1, structures=["box"])
+    """
+
+    # NOTE: this is basically a wrapper for backwards compatibility.
+
+    # encapsulating the log.warning since otherwise it is output everytime there is
```

Comment:
oh maybe this is why it's done in the init? would it be sufficient to just warn with `log_once=True`?

Reply:
Initially I had the log underneath the class, so it would issue the warning every time there was the `isinstance()` check. But I tried earlier with the root validator and it has the same effect as having it within the `__init__`, so I removed the init function and kept the root validator